Closing the write and the watch stream after 60s of idleness

schmidt-sebastian opened the pull request:
This is a port of https://github.com/firebase/firebase-ios-sdk/pull/388

jshcrowthe commented on packages/firestore/test/integration/util/helpers.ts:
```diff
@@ -23,6 +23,9 @@ import firebase from './firebase_export';
 import { PlatformSupport } from '../../../src/platform/platform';
 import { AsyncQueue } from '../../../src/util/async_queue';
+import { Firestore, FirestoreDatabase } from '../../../src/api/database';
+import { FirebaseApp } from '../../../../app/index';
```
This will break, if you are trying to reference other modules you need to do it directly

_i.e._
```javascript
import { FirebaseApp } from '@firebase/app';
```

schmidt-sebastian replied:
Done. Thanks for debugging this!

wilhuff commented on packages/firestore/src/remote/persistent_stream.ts:
```diff
@@ -253,6 +245,108 @@ export abstract class PersistentStream<
+  /**
+   * Closes the stream and cleans up as necessary:
+   *
+   * <ul>
```
Nit: Typescript comments aren't thinly veiled HTML the way Javadoc comments are. Do this as a markdown list, like we did in [mutation.ts](https://github.com/firebase/firebase-js-sdk/blob/deedb015d3bd7b63dca10459f87180a73e5c77a9/packages/firestore/src/model/mutation.ts#L333).

schmidt-sebastian replied:
Adjusted. Thanks for pointing this out.

wilhuff commented on packages/firestore/src/remote/persistent_stream.ts:
```diff
@@ -253,6 +245,108 @@ export abstract class PersistentStream<
+    error?: FirestoreError
+  ): Promise<void> {
+    assert(
+      finalState == PersistentStreamState.Error || isNullOrUndefined(error),
```
As in the iOS PR, this allows the error to be nil if the finalState is FSTStreamStateError. Is that desirable?

schmidt-sebastian replied:
This gets invoked by 'handleStreamClose' (the GRPC callback handler) and making this stricter would require that the server only ever closes a stream with an error.

wilhuff commented on packages/firestore/src/util/async_queue.ts:
```diff
@@ -39,24 +40,25 @@ export class AsyncQueue {
-      this.delayedOpCount++;
       const deferred = new Deferred<T>();
-      setTimeout(() => {
+      let handle = window.setTimeout(() => {
```
Why `window.setTimeout()` instead of the global? This is incompatible with node without extra hacks to define window.

schmidt-sebastian replied:
`setTimeout()` is defined to (perhaps incorrectly) return a Timer. See https://github.com/Microsoft/TypeScript/issues/842#issuecomment-58213196

wilhuff commented on packages/firestore/src/util/async_queue.ts:
```diff
@@ -39,24 +40,25 @@ export class AsyncQueue {
-      this.delayedOpCount++;
       const deferred = new Deferred<T>();
-      setTimeout(() => {
+      let handle = window.setTimeout(() => {
```
Also, this can be `const handle` instead of `let handle`

schmidt-sebastian replied:
Done

wilhuff commented on packages/firestore/src/util/async_queue.ts:
```diff
@@ -93,11 +95,18 @@ export class AsyncQueue {
+    this.delayedOperations.forEach((deferred, handle) => {
+      window.clearTimeout(handle);
+      deferred.reject(
+        new FirestoreError(Code.CANCELLED, 'Cancelled during queue drain')
```
These failures are potentially user visible so we need a better message about what can cause these to happen and how to avoid them if that's a problem.

If draining only happens during shutdown, the message can reflect that: `'Operation cancelled by shutdown'`.

schmidt-sebastian replied:
This is only used in tests, but I could see that we may want to call this during shutdown as well.

schmidt-sebastian commented on packages/firestore/src/api/database.ts:
```diff
@@ -455,6 +455,11 @@ export class Firestore implements firestore.Firestore, FirebaseService {
     }
   }
+
+  /** Creates a new AsyncQueue. Can be overwritten to provide a custom queue. */
```
Done

schmidt-sebastian commented on packages/firestore/src/remote/persistent_stream.ts:
```diff
@@ -253,6 +245,108 @@ export abstract class PersistentStream<
+        return this.handleIdleCloseTimer();
+      }, IDLE_TIMEOUT_MS)
+      .catch((err: FirestoreError) => {
+        // We ignore Promise rejections for cancelled idle checks.
```
The reason I added this is that the running the tests with this spewed a bunch of warnings in Chrome about unrejected Promises that were not handled. I made this comment better.

schmidt-sebastian commented:
@wilhuff I fixed the failing integration tests. The types we referred to in the tests ended up being minified, while the Public API that we checked against was not. I migrated the test to use the internal representation of the Public API. You can see a test run here: 
https://gist.github.com/schmidt-sebastian/18116a7c92ffbfdf1d535720850e63e8

schmidt-sebastian commented:
@mikelehen I addressed your feedback.

The drain() method in the AsyncQueue now optionally allows for immediate execution of delayed tasks. Furthermore, I have exposed this queue via firestore.INTERNAL. The tests can now access the public Firestore client and no longer need to rely on an internal client that exposes the queue.

mikelehen commented on packages/firestore/src/util/async_queue.ts:
```diff
@@ -17,16 +17,35 @@
+  // may also contain entries that have already been run (in which case `handle` is
+  // null).
+  //
+  // tslint:disable-next-line:no-any Accept any type of delayed operation.
```
Can you use AnyJs instead of "any" ?

schmidt-sebastian replied:
Done.

mikelehen commented on packages/firestore/src/util/async_queue.ts:
```diff
@@ -39,25 +58,37 @@ export class AsyncQueue {
-        this.delayedOpCount--; // decrement once it's actually queued.
+        this.delayedOperationsCount--;
+        if (this.delayedOperationsCount > 0) {
+          delayedOp.handle = null;
```
nit but I'm not a fan of this sort of optimization. It saves an assignment in some cases, but reads much less naturally (to me at least). I'd just do:
```
delayedOp.handle = null;
if (this.delayedOperationsCount === 0) {
  this.delayedOperations = [];
}
```

If you insist on optimizing this sort of thing, you should go ahead and optimize away the decrement too!
```
if (this.delayedOperationsCount !== 1) {
  this.delayedOperationsCount--;
  delayedOp.handle = null;
} else {
  this.delayedOperations = [];
}
```
(please don't actually do that :-))

schmidt-sebastian replied:
This wasn't meant to be the newest and greatest performance optimization. Rather, it just felt naturally to do this at the time. Your version is a bit more concise, so I went with that. Thanks.

mikelehen commented on packages/firestore/src/util/async_queue.ts:
```diff
@@ -39,25 +58,37 @@ export class AsyncQueue {
-      const deferred = new Deferred<T>();
-      setTimeout(() => {
+      this.delayedOperationsCount++;
+      const nextIndex = this.delayedOperations.length;
```
I think you can move this down and do:

const delayedOpIndex = this.delayedOperations.push(delayedOp);

schmidt-sebastian replied:
I don't actually need the index anymore and removed it altogether.

mikelehen commented on packages/firestore/src/util/async_queue.ts:
```diff
@@ -93,11 +124,31 @@ export class AsyncQueue {
-    return this.schedule(() => Promise.resolve(undefined));
+  /**
+   * Waits until all currently scheduled tasks are finished executing. Tasks
+   * schedule with a delay can be rejected or queued for immediate execution.
```
schedule => scheduled

schmidt-sebastian replied:
Done

mikelehen commented on packages/firestore/src/api/database.ts:
```diff
@@ -374,7 +375,8 @@ export class Firestore implements firestore.Firestore, FirebaseService {
     disableNetwork: () => this._firestoreClient.disableNetwork(),
-    enableNetwork: () => this._firestoreClient.enableNetwork()
+    enableNetwork: () => this._firestoreClient.enableNetwork(),
+    queue: this._queue
```
I'm not a big fan of exposing the whole queue externally since:

1) Although not documented or supported, this is now exposed to consumers of the SDK and could be abused.
2) The only reason this works is because our minification sucks right now and we only minify things that begin or end with `_`.  Really, the entire AsyncQueue class should be minified and all of the methods should have obfuscated names.
3) In general, I'd rather we make the testable API explicit and minimal. Exposing more than we need to leads to a slippery slope heading towards undesirable tight coupling between tests and implementation.

So I'd recommend exposing drainAsyncQueue() or similar (FWIW if you wanted you could make drain() always execute delayed tasks; I don't think it needs to be optional) and either remove the delayedOperationsCount assertions from tests or have the exposed method return the count of operations that were drained or something.

schmidt-sebastian replied:
I went with `drainAsyncTasks`. In the integration tests that use Firestore, I can indeed always execute the delayed tasks. I can't do this in the spec tests (since that always ends up closing the stream and confuses the tests).

The stream tests use Datastore directly and I allowed the queue to be passed in. That way, I can verify that I have at least one pending task.
